book/src/read-only-read-write-accounts.md

```diff
@@ -0,0 +1,135 @@
+read-only accounts are all grouped together at the end of the `account_keys`
+vector.  Counting read-only accounts from the end allows for the default `0`
+value to still be functionally correct, since a transaction will succeed with
+all read-write accounts.
+
+Since accounts can only appear once in the transaction's `account_keys` array,
+an account can only be read-only or read-write in a single transaction, not
+both.  The runtime treats a transaction as one atomic unit of execution. If any
+instruction needs read-write access to an Account a copy needs to be made the
+write lock needs to be held for the entire time the transaction is being
+processed by the runtime the runtime.
+
+## Starvation
+
+Read locks can keep the runtime from executing transactions requesting a write.
+
+When a request for a write lock is made while a read lock is open, the
+transaction requesting the write lock should be cached.  Upon closing the read
+lock the pending transactions can be pushed through the runtime.
+
+While a pending write transaction exists, any additional read lock requests for
+that account should fail.  It follows that any other write lock requests will
+fail as well.  Currently clients must retransmit when the transaction fails due
+to a pending transaction.  This approach would mimic that behavior as close as
+possible while preventing write starvation.
+
+## Program execution with read-only accounts
+
+Before handing of the accounts to program execution, the runtime can mark each
+account in each instruction as read-only. The read-only accounts can be passed
+as references without an extra copy.  The transaction will abort on a write to
+read-only.
+
+An alternative is to detect writes to read-only accounts and fail the
+transaction before commit.
+
+## Alternative design
+
+This design attempts to cache a read-only account after loading without the use
+of a transaction specified read-only accounts list.  Instead the read-only
+accounts are held in a reference counted table inside the runtime as the
+transactions are processed.
+
+1. Transaction accounts are locked.
+    a. If the account is present in the `read-only` table the TX does not fail.
+       The pending state for this TX is marked NeedReadLock.
+2. Transaction accounts are loaded.
+    a. Transaction accounts that are read-only increase their reference
+       count in the `read-only` table.
+    b. Transaction accounts that need a write lock and are present in the
+       `read-only` table fail.
+3. Transaction accounts are unlocked.
+    a. Decrement the `read-only` lock table reference count, remove if its 0
+    b. Remove from the `lock` set if the account is not in the `read-only`
+       table.
+
+The downside with this approach is that if the `lock` set mutex is released
+between lock and load to allow better pipelining of transactions, a request for
+a read-only account may fail, so this design is not suitable for treating
```

Comment:
This is a run-on sentence.  Maybe move this paragraph and the one below to a "Caveats" section?